nltk/tokenize/sonority_sequencing.py

```diff
@@ -0,0 +1,145 @@
+# Author: Christopher Hench <chris.l.hench@gmail.com>
+#         Alex Estes
+# URL: <http://nltk.sourceforge.net>
+# For license information, see LICENSE.TXT
+#
+# References: Otto Jespersen, Lehrbuch der Phonetik.
+#             University of Toronto (Leipzig, Teubner, 1904).
+#             Chapter 13, Silbe, pp. 185-203.
+#
+#             Elisabeth Selkirk, On the major class features and syllable theory. 
+#             In Aronoff & Oehrle (eds.) Language Sound Structure: Studies in Phonology.
+#             (Cambridge, MIT Press, 1984). pp. 107-136.
+#
+#             Susan Bartlett, et al. On the Syllabification of Phonemes.
+#             Human Language Technologies: The 2009 Annual Conference
+#             of the North American Chapter of the ACL.
+#             (Association for Computational Linguistics, 2009). pp. 308-316.
+
+"""
+The Sonority Sequencing Principle (SSP) is a language agnostic algorithm proposed
+by Otto Jesperson in 1904. The sonorous quality of a phoneme is judged by the
+openness of the lips. Syllable breaks occur before troughs in sonority. For more
+on the SSP see Selkirk (1984).
+
+The default implementation uses the English alphabet, but the `sonority_hiearchy`
+can be modified to IPA or any other alphabet for the use-case. The SSP is a
+universal syllabification algorithm, but that does not mean it performs equally
+across languages. Bartlett et al. (2009) is a good benchmark for English accuracy
+if utilizing IPA (pg. 311).
+
+Importantly, if a custom hiearchy is supplied and vowels span across more than
+one level, they should be given separately to the `vowels` class attribute.
+"""
+
+from __future__ import unicode_literals
+from nltk.tokenize.api import TokenizerI
+import re
+
+
+class SyllableTokenizer(TokenizerI):
+    '''
+    Syllabifies words based on the Sonority Sequencing Principle (SSP)
+
+    >>> SSP = SyllableTokenizer()
+    >>> SSP.tokenize('justification')
+    ['jus', 'ti', 'fi', 'ca', 'tion']
+    '''
+
+    def __init__(self, lang='en', sonority_hierarchy=False):
+        '''
+        Sonority hierarchy should be provided in descending order.
+        If vowels are spread across multiple levels, they should be
+        passed assigned self.vowels var together, otherwise should be
+        placed in first index of hierarchy.
+        '''
+
+        if not sonority_hierarchy and lang == 'en':
+            sonority_hierarchy = [
+                'aeiouy',  # vowels
+                'lmnrw',  # nasals
+                'zvsf',  # fricatives
+                'bcdgtkpqxhj'  # stops
+            ]
+
+        self.vowels = sonority_hierarchy[0]
+        self.phoneme_map = {}
+        for i, level in enumerate(sonority_hierarchy):
+            for c in level:
+                self.phoneme_map[c] = len(sonority_hierarchy) - i
+
+    def assign_values(self, token):
+        '''
+        Assigns each phoneme its value from the sonority hierarchy
+        '''
+        self.syllables_values = []
```

Comment:
Hmmm, this isn't a good use of `self`. Since for every token the `syllables_values` resets and the content changes, it doesn't belong to the `SyllableTokenizer` variable. Try returning `syllable_values` instead.

```python
def assign_values(self, token):
    syllable_values = []
    ...
    return syllable_values
```

It should just be `syllables_values` here without the self. And then in the `tokenize()` function, it should be:

```python
syllables_values = self.assign_values(token)
```